sw/device/lib/dif/dif_usbdev.c

```diff
@@ -0,0 +1,934 @@
+  uint32_t bytes_to_copy = usbdev->read_buffer_op_state.remaining_bytes;
+  if (bytes_to_copy > dst_len) {
+    bytes_to_copy = dst_len;
+  }
+  // Get the address of the buffer entry
+  uint32_t buffer_addr = USBDEV_BUFFER_REG_OFFSET +
+                         (usbdev->read_buffer_op_state.buffer_entry *
+                          USBDEV_BUFFER_ENTRY_SIZE_BYTES) +
+                         usbdev->read_buffer_op_state.offset;
+  // Copy payload to `dst`
+  mmio_region_memcpy_from_mmio32(usbdev->base_addr, buffer_addr, dst,
+                                 bytes_to_copy);
+  // Update current buffer operation
+  usbdev->read_buffer_op_state.offset += bytes_to_copy;
+  usbdev->read_buffer_op_state.remaining_bytes -= bytes_to_copy;
+  // Check if the entire payload is read
+  bool packet_done = false;
+  if (usbdev->read_buffer_op_state.remaining_bytes == 0) {
+    packet_done = true;
+    // Finish read buffer operation
+    buffer_op_stop_read(usbdev);
+    // Return the buffer to the free buffer pool
+    buffer_pool_add(&usbdev->buffer_pool,
+                    usbdev->read_buffer_op_state.buffer_entry);
+  }
+
+  if (bytes_written != NULL) {
+    *bytes_written = bytes_to_copy;
+  }
+
+  if (packet_done) {
+    return kDifUsbdevRxPacketReadBytesResultOK;
+  } else {
+    return kDifUsbdevRxPacketReadBytesResultContinue;
+  }
+}
+
+dif_usbdev_rx_packet_discard_bytes_result_t dif_usbdev_rx_packet_discard_bytes(
+    dif_usbdev_t *usbdev) {
+  if (usbdev == NULL) {
+    return kDifUsbdevRxPacketDiscardBytesResultInvalidArgument;
+  }
+
+  // Make sure that a read buffer operation is in progress
+  if (!buffer_op_is_reading(usbdev)) {
+    return kDifUsbdevRxPacketDiscardBytesResultNotReading;
+  }
+  // Finish read buffer operation
+  buffer_op_stop_read(usbdev);
+  // Return the buffer to the free buffer pool
+  buffer_pool_add(&usbdev->buffer_pool,
+                  usbdev->read_buffer_op_state.buffer_entry);
+
+  return kDifUsbdevRxPacketDiscardBytesResultOK;
+}
+
+dif_usbdev_tx_packet_write_bytes_result_t dif_usbdev_tx_packet_write_bytes(
+    dif_usbdev_t *usbdev, uint32_t endpoint, uint8_t *src, uint32_t src_len,
+    uint32_t *bytes_written) {
+  if (usbdev == NULL || (src_len > 0 && src == NULL)) {
+    return kDifUsbdevInPacketWriteBytesResultInvalidArgument;
+  }
+
+  if (!is_valid_endpoint(endpoint)) {
+    return kDifUsbdevInPacketWriteBytesResultGeneralError;
+  }
+
+  // Get tx packet status for the endpoint
+  dif_usbdev_tx_packet_status_t in_packet_status;
+  if (dif_usbdev_tx_packet_get_status(usbdev, endpoint, &in_packet_status) !=
+      kDifUsbdevTxPacketGetStatusResultOK) {
+    return kDifUsbdevInPacketWriteBytesResultGeneralError;
+  }
+
+  // HW details
+  const uint32_t config_in_reg_offset =
+      kEndpointHwInfos[endpoint].config_in_reg_offset;
+  uint32_t config_in_val;
+  uint32_t buffer_entry;
+  uint32_t buffer_len;
+
+  switch (in_packet_status) {
+    case kDifUsbdevTxPacketStatusNoPacket:
```

Comment:
Probably a good idea to wrap a big case in braces, just to avoid shenanigans with variable having to declare some variables outside of the `switch`.

Reply:
No variables are declared in these `case`s but still this is a good point. Done.